packages/lib/restrictionSchedule.ts

```diff
@@ -0,0 +1,97 @@
+  bookerTimezone?: string;
+}
+
+/**
+ * Checks if the restriction schedule feature is enabled for a team
+ * @param teamId - The team ID to check
+ * @returns Promise<boolean> - True if the feature is enabled, false otherwise
+ */
+export async function isRestrictionScheduleEnabled(teamId?: number): Promise<boolean> {
+  if (!teamId) {
+    return false; // Personal events don't have restriction schedules
+  }
+
+  const featureRepo = new FeaturesRepository();
+  return await featureRepo.checkIfTeamHasFeature(teamId, "restriction-schedule");
+}
+/**
+ * Checks if a booking time falls within the restriction schedule's availability
+ * @param params - The restriction schedule check parameters
+ * @returns true if the booking is allowed, false otherwise
+ * @throws Error if required timezone information is missing
+ */
+export function isBookingAllowedByRestrictionSchedule({
+  restrictionSchedule,
+  bookingStartTime,
+  bookingEndTime,
+  useBookerTimezone,
+  bookerTimezone,
+}: RestrictionScheduleCheckParams): boolean {
+  if (!restrictionSchedule.timeZone && !useBookerTimezone) {
+    throw new Error("No timezone is set for the restriction schedule and useBookerTimezone is false");
+  }
+
+  if (useBookerTimezone && !bookerTimezone) {
+    throw new Error("Booker timezone is required when useBookerTimezone is true");
+  }
+
+  const restrictionTimezone = useBookerTimezone ? bookerTimezone! : restrictionSchedule.timeZone!;
+
+  const dateOverrides = restrictionSchedule.availability.filter((a) => !!a.date);
+  const recurringRules = restrictionSchedule.availability.filter((a) => !a.date);
+
+  const bookingStartTimeNormalized = dayjs(bookingStartTime.utc()).tz(restrictionTimezone);
+  const bookingEndTimeNormalized = dayjs(bookingEndTime.utc()).tz(restrictionTimezone);
+  const bookingDateStr = bookingStartTimeNormalized.format("YYYY-MM-DD");
+  const bookingWeekday = bookingStartTimeNormalized.day();
+  const bookingStartValue = bookingStartTime.valueOf();
+  const bookingEndValue = bookingEndTime.valueOf();
+
+  const overrideRulesForTheDay = dateOverrides.filter((a) =>
+    dayjs.utc(a.date).isSame(bookingStartTime.utc(), "day")
+  );
+  const recurringRulesForTheDay = recurringRules.filter((a) => a.days.includes(bookingWeekday));
+
+  if (overrideRulesForTheDay.length > 0) {
+    const overrideAllowsBooking = overrideRulesForTheDay.some((rule) => {
+      const startTimeStr = dayjs.utc(rule.startTime).format("HH:mm");
+      const endTimeStr = dayjs.utc(rule.endTime).format("HH:mm");
+      const start = dayjs.tz(`${bookingDateStr}T${startTimeStr}`, restrictionTimezone);
+      const end = dayjs.tz(`${bookingDateStr}T${endTimeStr}`, restrictionTimezone);
+      return bookingStartValue >= start.valueOf() && bookingEndValue <= end.valueOf();
+    });
+    return overrideAllowsBooking;
+  }
+  if (recurringRulesForTheDay.length === 0) return false;
+  const recurringAllowsBooking = recurringRulesForTheDay.some((rule) => {
```

Comment:
same here, we can't do it like that and need to work with `dateRanges` instead. This will not cover slots that go over midnight

Example: 
restriction schedule 11:45PM-12:15AM
event duration: 30 min
Bug: this slot won't show as available